hypershift-operator/controllers/nodepool/nto.go

```diff
@@ -0,0 +1,406 @@
+	if err := cg.List(ctx, nodeTuningGeneratedConfigs, client.MatchingLabels{
+		nodeTuningGeneratedConfigLabel: "true",
+		hyperv1.NodePoolLabel:          cg.nodePool.GetName(),
+	}, client.InNamespace(cg.controlplaneNamespace)); err != nil {
+		return nil, err
+	}
+	return nodeTuningGeneratedConfigs.Items, nil
+}
+
+// BuildMirrorConfigs returns a slice of MirrorConfigs for all the NTO generated config.
+func BuildMirrorConfigs(ctx context.Context, cg *ConfigGenerator) ([]*MirrorConfig, error) {
+	var configs []corev1.ConfigMap
+	// Look for NTO generated MachineConfigs from the hosted control plane namespace
+	nodeTuningGeneratedConfigs, err := getNTOGeneratedConfig(ctx, cg)
+	if err != nil {
+		return nil, err
+	}
+	configs = append(configs, nodeTuningGeneratedConfigs...)
+
+	userConfig, err := cg.getUserConfigs(ctx)
+	if err != nil {
+		return nil, err
+	}
+	configs = append(configs, userConfig...)
+
+	var errors []error
+	var mirrorConfigs []*MirrorConfig
+	for i, config := range configs {
+		cmPayload := config.Data[TokenSecretConfigKey]
+		// ignition config-map payload may contain multiple manifests
+		yamlReader := yaml.NewYAMLReader(bufio.NewReader(strings.NewReader(cmPayload)))
+		for {
+			manifestRaw, err := yamlReader.Read()
+			if err != nil && err != io.EOF {
+				errors = append(errors, fmt.Errorf("configmap %q contains invalid yaml: %w", config.Name, err))
+				continue
+			}
+			if len(manifestRaw) != 0 && strings.TrimSpace(string(manifestRaw)) != "" {
+				mirrorConfig, err := getMirrorConfigForManifest(manifestRaw, cg)
+				if err != nil {
+					errors = append(errors, fmt.Errorf("configmap %q yaml document failed validation: %w", config.Name, err))
+					continue
+				}
+				if mirrorConfig != nil {
+					mirrorConfig.ConfigMap = &configs[i]
+					mirrorConfigs = append(mirrorConfigs, mirrorConfig)
+				}
+			}
+			if err == io.EOF {
+				break
+			}
+		}
+	}
+
+	return mirrorConfigs, utilerrors.NewAggregate(errors)
+}
+
+// getMirrorConfigForManifest returns a MirrorConfig for a ContainerRuntimeConfig manifest or nil.
+func getMirrorConfigForManifest(manifest []byte, cg *ConfigGenerator) (*MirrorConfig, error) {
+	scheme := runtime.NewScheme()
+	_ = mcfgv1.Install(scheme)
+
+	yamlSerializer := serializer.NewSerializerWithOptions(
+		serializer.DefaultMetaFactory, scheme, scheme,
+		serializer.SerializerOptions{Yaml: true, Pretty: true, Strict: false},
+	)
+	// for manifests that should be mirrored into hosted control plane namespace
+	var mirrorConfig *MirrorConfig
+	cr, _, err := yamlSerializer.Decode(manifest, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding config: %w", err)
+	}
+
+	switch obj := cr.(type) {
+	case *mcfgv1.ContainerRuntimeConfig:
+		obj.Spec.MachineConfigPoolSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"machineconfiguration.openshift.io/mco-built-in": "",
+			},
+		}
+		_, err = encode(cr, yamlSerializer)
```

Comment:
I don't understand this part, why are we decoding the object, changing it, encoding again but not use the result?

Reply:
seems like this function is effectively always returning 
```go
return &MirrorConfig{Labels: map[string]string{ContainerRuntimeConfigConfigMapLabel: ""}}
```

Reply:
note for reviewer: this function is a result of trying to keep the nto stuff decoupled for now.
you're right, that bit is a carry over from the original coupled func and it's not needed, let me update.